examples/mlmc.cpp

```diff
@@ -0,0 +1,351 @@
+    mfem::Vector rhs_u_fine;
+
+    // Setting up finite volume discretization problem
+    SPE10Problem spe10problem(permFile, nDimensions, spe10_scale, slice,
+                              metis_agglomeration, coarseningFactor);
+
+    mfem::ParMesh* pmesh = spe10problem.GetParMesh();
+
+    if (myid == 0)
+    {
+        std::cout << pmesh->GetNEdges() << " fine edges, " <<
+                  pmesh->GetNFaces() << " fine faces, " <<
+                  pmesh->GetNE() << " fine elements\n";
+    }
+
+    ess_attr.SetSize(nbdr);
+    for (int i(0); i < nbdr; ++i)
+        ess_attr[i] = ess_zeros[i];
+
+    // Construct "finite volume mass" matrix using mfem
+    mfem::RT_FECollection sigmafec(0, nDimensions);
+    mfem::ParFiniteElementSpace sigmafespace(pmesh, &sigmafec);
+
+    {
+        mfem::ParBilinearForm a(&sigmafespace);
+        a.AddDomainIntegrator(
+            new FiniteVolumeMassIntegrator(*spe10problem.GetKInv()) );
+        a.Assemble();
+        a.Finalize();
+        a.SpMat().GetDiag(weight);
+    }
+
+    for (int i = 0; i < weight.Size(); ++i)
+    {
+        weight[i] = 1.0 / weight[i];
+    }
+
+    mfem::L2_FECollection ufec(0, nDimensions);
+    mfem::ParFiniteElementSpace ufespace(pmesh, &ufec);
+
+    mfem::LinearForm q(&ufespace);
+    q.AddDomainIntegrator(
+        new mfem::DomainLFIntegrator(*spe10problem.GetForceCoeff()) );
+    q.Assemble();
+    rhs_u_fine = q;
+
+    // Construct vertex_edge table in mfem::SparseMatrix format
+    auto& vertex_edge_table = nDimensions == 2 ? pmesh->ElementToEdgeTable()
+                              : pmesh->ElementToFaceTable();
+    mfem::SparseMatrix vertex_edge = TableToMatrix(vertex_edge_table);
+
+    // Construct agglomerated topology based on METIS or Cartesion aggloemration
+    mfem::Array<int> partitioning;
+    if (metis_agglomeration)
+    {
+        MetisPart(partitioning, sigmafespace, ufespace, coarseningFactor);
+    }
+    else
+    {
+        auto num_procs_xyz = spe10problem.GetNumProcsXYZ();
+        CartPart(partitioning, num_procs_xyz, *pmesh, coarseningFactor);
+    }
+
+    const auto& edge_d_td(sigmafespace.Dof_TrueDof_Matrix());
+
+    auto edge_boundary_att = GenerateBoundaryAttributeTable(pmesh);
+
+    // Create Upscaler and Solve
+    FiniteVolumeMLMC fvupscale(comm, vertex_edge, weight, partitioning, *edge_d_td,
+                               edge_boundary_att, ess_attr, spect_tol, max_evects);
+
+    mfem::Array<int> marker(fvupscale.GetFineMatrix().getD().Width());
+    marker = 0;
+    sigmafespace.GetEssentialVDofs(ess_attr, marker);
+
+    fvupscale.PrintInfo();
+    fvupscale.ShowSetupTime();
+
+    mfem::BlockVector rhs_fine(fvupscale.GetFineBlockVector());
+    rhs_fine.GetBlock(0) = 0.0;
+    rhs_fine.GetBlock(1) = rhs_u_fine;
+
+    const int num_fine_edges = vertex_edge.Width();
+    const int num_aggs = partitioning.Max() + 1; // this can be wrong if there are empty partitions
+    StupidSampler sampler(num_fine_edges, num_aggs);
```

Comment:
@barker29 this is a great name! Do you want to keep this name or use something like `SimpleSampler`?

Reply:
Yes, that's more professional. Done.